File: src/statistics/chi2_context.py
"""Test whether flanking bases significantly affect the central-base substitution rate.

Input:
    one TSV file (output of trisbst_2TSVs.py)

Output:
    for each of the 6 central mutation types (C>A, C>G, C>T, T>A, T>C, T>G):
        chi-square statistic, degrees of freedom, p-value, eta-squared

Method: chi-square goodness-of-fit against a uniform-rate null hypothesis.
For each central mutation type m (e.g. C>A) and its flanking contexts i:
    r_hat[m] = sum_i(mutNum[m,i]) / sum_i(rootNum[m,i])
    E[m,i]   = rootNum[m,i] * r_hat[m]
    chi2[m]  = sum_i (mutNum[m,i] - E[m,i])^2 / E[m,i]
    df       = (number of valid contexts) - 1

Using rootNum as exposure accounts for empirical trinucleotide frequencies.
Bonferroni correction applies across the 6 central mutation types (threshold p < 0.05/6).
"""

import argparse
import csv
import math

from scipy.stats import chi2 as chi2_dist

BONFERRONI_N = 6  # number of central mutation types


def load_tsv(path):
    """Return {mutType: (mutNum, totalRootNum)} from a trinucleotide TSV file."""
    data = {}
    with open(path) as f:
        reader = csv.DictReader(f, delimiter="\t")
        for row in reader:
            data[row["mutType"]] = (float(row["mutNum"]), float(row["totalRootNum"]))
    return data


def central_mut(mut_type):
    """Extract central mutation type from 'X[Y>Z]W' format. Returns 'Y>Z'."""
    return mut_type[2:5]


def group_by_central(data):
    """Group TSV data by central mutation type.

    Returns {central_mut: {mutType: (mutNum, totalRootNum)}}.
    """
    groups = {}
    for mut_type, counts in data.items():
        cm = central_mut(mut_type)
        if cm not in groups:
            groups[cm] = {}
        groups[cm][mut_type] = counts
    return groups


def compute_chi2_context(contexts):
    """Compute chi-square goodness-of-fit for one central mutation type.

    contexts: {mutType: (mutNum, totalRootNum)}
    Returns (chi2_stat, df, p_value, eta2).
    """
    total_mut = sum(m for m, _ in contexts.values())
    total_root = sum(r for _, r in contexts.values())

    if total_root == 0.0 or total_mut == 0.0:
        return float("nan"), 0, float("nan"), float("nan")

    r_hat = total_mut / total_root

    chi2_stat = 0.0
    valid = 0
    for mut_num, root_num in contexts.values():
        if root_num == 0.0:
            continue
        exp = root_num * r_hat
        if exp == 0.0:
            continue
        chi2_stat += (mut_num - exp) ** 2 / exp
        valid += 1

    df = max(valid - 1, 0)
    p_value = chi2_dist.sf(chi2_stat, df) if df > 0 else float("nan")
    eta2 = chi2_stat / (chi2_stat + total_mut) if (chi2_stat + total_mut) > 0.0 else float("nan")

    return chi2_stat, df, p_value, eta2


if __name__ == "__main__":
    parser = argparse.ArgumentParser(
        description=(
            "Chi-square goodness-of-fit test for context dependence of substitution "
            "rates. Tests whether flanking bases significantly affect the substitution "
            "rate for each central mutation type."
        )
    )
    parser.add_argument("tsv", help="TSV file (output of trisbst_2TSVs.py)")
    parser.add_argument("--label", default="", help="dataset label for display")
    args = parser.parse_args()

    data = load_tsv(args.tsv)
    groups = group_by_central(data)

    if args.label:
        print(args.label)

    threshold = 0.05 / BONFERRONI_N
    header = (
        f"{'mutType':<8}  {'chi2':>12}  {'df':>4}  {'p':>12}  {'eta2':>8}  sig*"
    )
    print(header)
    print("-" * len(header))

    for cm in sorted(groups):
        chi2_stat, df, p_value, eta2 = compute_chi2_context(groups[cm])
        sig = "*" if (not math.isnan(p_value) and p_value < threshold) else ""
        p_str = f"{p_value:.4e}" if not math.isnan(p_value) else "nan"
        eta2_str = f"{eta2:.4f}" if not math.isnan(eta2) else "nan"
        print(
            f"{cm:<8}  {chi2_stat:>12.4f}  {df:>4}  {p_str:>12}  {eta2_str:>8}  {sig}"
        )

    print(f"\n* Bonferroni threshold: p < {threshold:.4f} (n = {BONFERRONI_N} tests)")

File: src/statistics/chi2_context.R
# Chi-square test for context dependence of substitution rates.
#
# For each of 6 central mutation types (C>A, C>G, C>T, T>A, T>C, T>G),
# tests whether the substitution rate is uniform across 16 flanking contexts.
# Falls back to Fisher's exact test (Monte Carlo) when expected counts < 5.
#
# Usage: Rscript chi2_context.R <tsv_file> [label]

args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 1) {
  stop("Usage: Rscript chi2_context.R <tsv_file> [label]")
}

tsv_path <- args[1]
label <- if (length(args) >= 2) args[2] else ""

data <- read.delim(tsv_path, header = TRUE, sep = "\t")

# Extract central mutation type: "X[Y>Z]W" -> "Y>Z"
data$central <- substr(data$mutType, 3, 5)

# Keep only rows with totalRootNum > 0
data <- data[data$totalRootNum > 0, ]

if (nrow(data) == 0) {
  stop("No rows with totalRootNum > 0")
}

central_types <- sort(unique(data$central))

if (label != "") {
  cat(label, "\n")
}
cat("* p < 0.05, ** p < 0.01\n\n")

cat(sprintf("%-8s  %-8s  %12s  %4s  %12s  %s\n",
            "mutType", "method", "statistic", "df", "p_value", "sig"))
cat(paste(rep("-", 56), collapse = ""), "\n")

for (ct in central_types) {
  sub <- data[data$central == ct, ]

  # Build 2 x N contingency table: row 1 = mutNum, row 2 = totalRootNum - mutNum
  mut_counts <- sub$mutNum
  non_counts <- sub$totalRootNum - sub$mutNum
  mat <- rbind(mut_counts, non_counts)

  # Remove columns where both rows are zero
  keep <- colSums(mat) > 0
  mat <- mat[, keep, drop = FALSE]

  if (ncol(mat) < 2) {
    cat(sprintf("%-8s  %-8s  %12s  %4s  %12s  %s\n",
                ct, "skip", "NA", "NA", "NA", ""))
    next
  }

  # Try chi-square test; check for low expected counts
  exp_counts <- suppressWarnings(chisq.test(mat))$expected
  has_low_expected <- any(exp_counts < 5)

  if (!has_low_expected) {
    result <- chisq.test(mat)
    method <- "chisq"
    stat <- result$statistic
    df <- result$parameter
    pval <- result$p.value
    stdres <- result$stdres[1, ]  # standardized residuals for mut row
    expected <- result$expected[1, ]  # expected counts for mut row
  } else {
    result <- fisher.test(mat, simulate.p.value = TRUE, B = 10000)
    method <- "fisher"
    stat <- NA
    df <- NA
    pval <- result$p.value
    # Compute stdres and expected manually for fisher fallback
    chi_result <- suppressWarnings(chisq.test(mat))
    stdres <- chi_result$stdres[1, ]
    expected <- chi_result$expected[1, ]
  }

  sig <- if (pval < 0.01) "**" else if (pval < 0.05) "*" else ""

  stat_str <- if (is.na(stat)) sprintf("%12s", "NA") else sprintf("%12.4f", stat)
  df_str <- if (is.na(df)) sprintf("%4s", "NA") else sprintf("%4d", df)

  cat(sprintf("%-8s  %-8s  %s  %s  %12.4e  %s\n",
              ct, method, stat_str, df_str, pval, sig))

  # Print per-context expected counts and standardized residuals
  context_names <- sub$mutType[keep]
  cat(sprintf("  %-12s  %8s  %8s\n", "context", "expected", "stdres"))
  for (j in seq_along(stdres)) {
    cat(sprintf("  %-12s  %8.2f  %8.2f\n", context_names[j], expected[j], stdres[j]))
  }
  cat("\n")
}
